File: src/commands/http.ts
import { type App } from '@/main/app.js'
import { isSilent } from '@/utils/cmd.js'
import { InvalidParamError, MissingParamError } from '@/utils/errors.js'
import { convertToJSON } from '@/utils/mappers.js'
import axios, { type AxiosResponse } from 'axios'
import color from 'picocolors'
import { log, outro } from '@clack/prompts'

type Method = 'get' | 'post' | 'put' | 'delete'

type Http = Record<
  Method,
  (
    url: string,
    headers: Record<string, string>,
    body: Record<string, unknown>,
  ) => Promise<AxiosResponse<unknown>>
>

async function httpCommand(params: string[]): Promise<void> {
  if (params.length === 0) {
    throw new MissingParamError('params')
  }

  const [method, urlParams] = getMethod(params)
  const [url, bodyAndHeadersParams] = getUrl(urlParams)
  const [bodyParams, headerParams] =
    splitBodyAndHeaderParams(bodyAndHeadersParams)
  const body = getBody(bodyParams)
  const headers = getHeaders(headerParams)

  const http = createHttp()
  const { status, data } = await http[method](url, headers, body)

  if (!isSilent()) {
    const statusLabel = color.cyan('"StatusCode"')
    const bodyLabel = color.cyan('"Body"')

    const statusMessage = `${statusLabel}: ${status}`
    status < 400 ? log.success(statusMessage) : log.error(statusMessage)

    if (data) {
      const jsonLines = JSON.stringify(data, null, 2).split(/\n/g)

      log.message(`${bodyLabel}: ${jsonLines[0]}`)

      for (let i = 1; i < jsonLines.length; i++) {
        process.stdout.moveCursor(0, -1)
        log.message(
          jsonLines[i].replace(/^(.*?)(".+?")/, `$1${color.cyan('$2')}`),
        )
      }
    } else {
      log.message(`${bodyLabel}: No Content`)
    }
    process.stdout.moveCursor(0, -1)
    outro()
  }
}

function getMethod(params: string[]): [Method, string[]] {
  const methods: Method[] = ['get', 'post', 'put', 'delete']
  const param = params[0].toLowerCase()

  if (methods.includes(param)) {
    return [param as Method, params.slice(1)]
  }

  return ['get', params]
}

function getUrl(params: string[]): [string, string[]] {
  const rest = params.slice(1)

  const firstParam = params[0]

  if (firstParam.startsWith('/')) {
    return [`http://localhost:3000${firstParam}`, rest]
  }

  if (firstParam.startsWith(':')) {
    return [`http://localhost${firstParam}`, rest]
  }

  if (firstParam.startsWith('http')) {
    return [firstParam, rest]
  }

  throw new InvalidParamError('url')
}

function getBody(bodyParams: string[]): Record<string, unknown> {
  return bodyParams.length > 0 ? convertToJSON(bodyParams) : {}
}

function getHeaders(params: string[]): Record<string, string> {
  return (
    params.length > 0
      ? convertToJSON(params.map(p => p.replace(/^h\./, '')))
      : {}
  ) as Record<string, string>
}

function splitBodyAndHeaderParams(params: string[]): [string[], string[]] {
  const bodyParams: string[] = []
  const headersParams: string[] = []

  for (const param of params) {
    if (param.startsWith('h.')) {
      headersParams.push(param)
    } else {
      bodyParams.push(param)
    }
  }

  return [bodyParams, headersParams]
}

function createHttp(): Http {
  const a = axios.create({
    validateStatus: () => true,
  })

  return {
    async get(url, headers): Promise<AxiosResponse> {
      return a.get(url, { headers })
    },
    async post(url, headers, body): Promise<AxiosResponse> {
      return a.post(url, body, { headers })
    },
    async put(url, headers, body): Promise<AxiosResponse> {
      return a.put(url, body, { headers })
    },
    async delete(url, headers): Promise<AxiosResponse> {
      return a.delete(url, { headers })
    },
  }
}

export function httpRecord(app: App): void {
  app.register({
    name: 'http',
    alias: null,
    params: ['<method?> <url> <body?> <headers?>'],
    description: 'Make an http request',
    example:
      'my http post /user key1=1 key2.subset1=true key2.subset2=3.14 key3=Hello+World h.authorization=token',
    action: httpCommand,
  })
}

File: cmd/http.go
/*
Copyright © 2024 NAME HERE <EMAIL ADDRESS>
*/
package cmd

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"regexp"
	"strconv"
	"strings"

	"github.com/Mist3rBru/go-clack/prompts"
	"github.com/Mist3rBru/go-clack/prompts/utils"
	"github.com/Mist3rBru/go-clack/third_party/picocolors"
	"github.com/spf13/cobra"
)

// httpCmd represents the http command
var httpCmd = &cobra.Command{
	Use:   "http",
	Short: "Make an http request",
	Long:  "Make an http request",
	Run: func(cmd *cobra.Command, args []string) {
		verifyArgs("params", args)

		method, urlParams := getMethod(args)
		uri, bodyAndHeadersParams, err := getUrl(urlParams)
		if err != nil {
			prompts.Error(err.Error())
			return
		}

		bodyParams, headerParams := splitBodyAndHeaderParams(bodyAndHeadersParams)
		headers := getHeaders(headerParams)
		body := getBody(bodyParams)

		request := &http.Request{
			Method: method,
			URL:    uri,
			Header: headers,
			Body:   body,
		}
		client := &http.Client{}
		response, err := client.Do(request)
		if err != nil {
			prompts.Error(err.Error())
			return
		}
		defer response.Body.Close()

		statusLabel := picocolors.Cyan("\"StatusCode\"")
		bodyLabel := picocolors.Cyan("\"Body\"")

		statusMessage := fmt.Sprintf("%s: %d", statusLabel, response.StatusCode)
		if response.StatusCode < 400 {
			prompts.Success(statusMessage)
		} else {
			prompts.Error(statusMessage)
		}

		responseJsonLines, err := parseJsonResponse(response)
		if err != nil {
			prompts.Error(err.Error())
			return
		}

		if len(responseJsonLines) > 0 {
			os.Stdout.WriteString(fmt.Sprintf("%s %s: %s\n", picocolors.Gray(utils.S_BAR), bodyLabel, responseJsonLines[0]))
			lineTemplate := fmt.Sprintf(`$1%s`, picocolors.Cyan("$2"))
			lineRegex := regexp.MustCompile(`^(.*?)(".+?")`)
			for i := 1; i < len(responseJsonLines); i++ {
				formattedLine := lineRegex.ReplaceAllString(responseJsonLines[i], lineTemplate)
				os.Stdout.WriteString(fmt.Sprintf("%s %s\n", picocolors.Gray(utils.S_BAR), formattedLine))
			}
		} else {
			os.Stdout.WriteString(fmt.Sprintf("%s %s: No Content\n", picocolors.Gray(utils.S_BAR), bodyLabel))
		}

		os.Stdout.WriteString(picocolors.Gray(utils.S_BAR_END) + "\n")
	},
}

func getMethod(args []string) (string, []string) {
	verifyArgs("method param", args)

	method := strings.ToUpper(args[0])
	switch method {
	case "GET", "POST", "PUT", "DELETE":
		return method, args[1:]
	}
	return "GET", args
}

func getUrl(args []string) (*url.URL, []string, error) {
	verifyArgs("url param", args)

	firstArg := args[0]
	rest := args[1:]

	if strings.HasPrefix(firstArg, "/") {
		uri, err := url.Parse(fmt.Sprintf("http://localhost:3000%s", firstArg))
		return uri, rest, err
	}

	if strings.HasPrefix(firstArg, ":") {
		uri, err := url.Parse(fmt.Sprintf("http://localhost%s", firstArg))
		return uri, rest, err
	}

	if strings.HasPrefix(firstArg, "http") {
		uri, err := url.Parse(firstArg)
		return uri, rest, err
	}

	return &url.URL{}, []string{}, fmt.Errorf("invalid url param")
}

func splitBodyAndHeaderParams(args []string) ([]string, []string) {
	var bodyParams, headersParams []string

	for _, arg := range args {
		if strings.HasPrefix(arg, "h.") {
			headersParams = append(headersParams, arg)
		} else {
			bodyParams = append(bodyParams, arg)
		}
	}

	return bodyParams, headersParams
}

func getBody(args []string) io.ReadCloser {
	if len(args) == 0 {
		return nil
	}
	bodyObj := convertToJson(args)
	bodyBytes, _ := json.Marshal(bodyObj)
	bodyBuffer := bytes.NewBuffer(bodyBytes)
	return io.NopCloser(bodyBuffer)
}

func getHeaders(args []string) http.Header {
	headers := http.Header{}
	headers.Add("Content-Type", "application/json")

	if len(args) == 0 {
		return headers
	}

	headerArgs := []string{}
	for _, arg := range args {
		headerArgs = append(headerArgs, strings.Replace(arg, "h.", "", 1))
	}

	headerObj := convertToJson(headerArgs)
	for header, value := range headerObj {
		headers.Add(header, value.(string))
	}

	return headers
}

func convertToJson(keyValueList []string) map[string]any {
	result := make(map[string]any)

	for _, item := range keyValueList {
		parts := strings.SplitN(item, "=", 2)
		keys := strings.Split(parts[0], ".")
		value := parts[1]

		currentObject := result

		for i := 0; i < len(keys)-1; i++ {
			key := keys[i]

			if _, exists := currentObject[key].(map[string]any); !exists {
				currentObject[key] = make(map[string]any)
			}
			currentObject = currentObject[key].(map[string]any)
		}

		currentObject[keys[len(keys)-1]] = parseValue(value)
	}

	return result
}

func parseValue(value string) any {
	if value == "" {
		return value
	} else if value == "true" {
		return true
	} else if value == "false" {
		return false
	} else if num, err := strconv.ParseFloat(value, 64); err == nil {
		return num
	} else if isArray(value) {
		elements := strings.Split(value[1:len(value)-1], ",")
		var result []any
		for _, element := range elements {
			if trimmedElement := strings.TrimSpace(element); trimmedElement != "" {
				result = append(result, parseValue(trimmedElement))
			}
		}
		return result
	} else if isObject(value) {
		var result map[string]any
		if err := json.Unmarshal([]byte(value), &result); err == nil {
			return result
		}
	}

	return strings.ReplaceAll(strings.Trim(value, `"'`), "+", " ")
}

func isArray(value string) bool {
	matched, _ := regexp.MatchString(`^\[.*?\]$`, value)
	return matched
}

func isObject(value string) bool {
	matched, _ := regexp.MatchString(`^{.*?}$`, value)
	return matched
}

func parseJsonResponse(response *http.Response) ([]string, error) {
	body, err := io.ReadAll(response.Body)
	if err != nil {
		return nil, err
	}

	var bodyJsonBuffer bytes.Buffer
	if err := json.Indent(&bodyJsonBuffer, body, "", "  "); err != nil {
		return nil, nil
	}

	bodyJsonLines := strings.Split(bodyJsonBuffer.String(), "\n")

	return bodyJsonLines, nil
}

func verifyArgs(label string, args []string) {
	if len(args) == 0 {
		prompts.Error(fmt.Sprintf("Missing %s", label))
		os.Exit(0)
	}
}

func init() {
	rootCmd.AddCommand(httpCmd)

	// Here you will define your flags and configuration settings.

	// Cobra supports Persistent Flags which will work for this command
	// and all subcommands, e.g.:
	// httpCmd.PersistentFlags().String("foo", "", "A help for foo")

	// Cobra supports local flags which will only run when this command
	// is called directly, e.g.:
	// httpCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
}
